Add support for Vertex AI model prefix format

changjonathanc opened the pull request:
# Add support for Vertex AI model prefix format

## Overview
This PR adds support for the `vertex_ai/` model name prefix format while preserving the original model name format for proper cost calculation.

## Problem
When using LiteLLM with Vertex AI's hosted Claude models, LiteLLM doesn't recognize the `vertex_ai/` prefix when passed in the format `vertex_ai/claude-3-7-sonnet@20250219`.

## Solution
This PR adds explicit handling for the `vertex_ai/` prefix in `get_llm_provider()` function to correctly identify and handle models with this prefix.

## Changes
- Updated `get_llm_provider()` in `litellm_core_utils/get_llm_provider_logic.py` to handle the `vertex_ai/` prefix
- Added a helper function `_format_model_name_for_cost_calc()` in `cost_calculator.py` that preserves the original model name format
- Added comprehensive tests in `tests/local_testing/test_vertex_ai_provider.py`

## Tests
Added a new test file `tests/local_testing/test_vertex_ai_provider.py` that tests:
- Handling of `vertex_ai/` prefix in model names
- Preservation of model name format for cost calculation
- End-to-end cost calculation for Vertex AI hosted Claude models

## Usage
With this PR, users can specify Vertex AI hosted models like this:
```python
import litellm

response = litellm.completion(
    model="vertex_ai/claude-3-7-sonnet@20250219",
    messages=[{"role": "user", "content": "Hello\!"}]
)
```

cccntu commented:
# Detailed Explanation of the Issue and Fix

## Root Issue

There are two connected issues with Vertex AI model handling:

### 1. Missing Provider Detection for `vertex_ai/` Prefix

When LiteLLM receives a model name with the `vertex_ai/` prefix (e.g., `vertex_ai/claude-3-7-sonnet@20250219`), the `get_llm_provider()` function doesn't have explicit handling for this format. This contrasts with other providers like `azure/` that have special handling.

The issue manifests during cost calculation in `cost_calculator.py` when the function tries to infer the provider from a model name without the prefix:

```python
# In cost_calculator.py, line ~201
_, custom_llm_provider, _, _ = litellm.get_llm_provider(model=model)
```

This fails with: `LLM Provider NOT provided. Pass in the LLM provider you are trying to call.`

### 2. Date Format Handling

Vertex AI uses the format `@<date>` (e.g., `claude-3-7-sonnet@20250219`) while direct Anthropic models use `-<date>` (e.g., `claude-3-7-sonnet-20250219`). Although the pricing database already has entries with the `@` format, additional model name formatting during cost calculation can sometimes cause format mismatches.

## Fix Implementation

Our PR addresses both issues:

1. **Added explicit handling for `vertex_ai/` prefix** in `get_llm_provider()`, ensuring it correctly identifies Vertex AI models:
   ```python
   if model.split("/", 1)[0].lower() == "vertex_ai":
       base_model = model.split("/", 1)[1]
       custom_llm_provider = "vertex_ai"
       return base_model, custom_llm_provider, dynamic_api_key, api_base
   ```

2. **Added a helper function** `_format_model_name_for_cost_calc()` to standardize model name formatting during cost calculation, ensuring proper matching in the pricing database:
   ```python
   if model_name and custom_llm_provider == "vertex_ai" and "@" in model_name:
       # Convert @ to - in date format to make it compatible with Anthropic models in cost DB
       return model_name.replace("@", "-")
   ```

## Testing

The PR includes comprehensive tests that verify:
1. The `vertex_ai/` prefix is correctly handled
2. Cost calculation works properly with various Vertex AI models (Claude, Llama, etc.)
3. End-to-end usage scenarios

This fix allows seamless usage of Vertex AI models in applications like SWE-agent without requiring custom monkey patching or workarounds.

## Example Usage

```python
# Now works correctly with both cost tracking and API calls
response = litellm.completion(
    model="vertex_ai/claude-3-7-sonnet@20250219", 
    messages=[{"role": "user", "content": "Hello!"}]
)
```

cccntu commented:
**Updated PR with Improved Model Name Handling**

I've updated this PR with a critical fix:

1. The original implementation was incorrectly converting @ to - in model names (e.g., claude-3-7-sonnet@20250219 → claude-3-7-sonnet-20250219).

2. Upon analyzing the pricing database, I discovered it already contains entries with the original @ format for Vertex AI models (e.g., vertex_ai/claude-3-sonnet@20240229).

3. The updated implementation:
   - Still properly identifies the vertex_ai/ prefix in get_llm_provider()
   - No longer transforms model names, preserving the @ format for accurate pricing database lookups
   - Includes updated tests that verify this behavior

This fix will ensure proper cost calculation for Vertex AI models while maintaining API compatibility.

cccntu commented:
**Further Simplified the Implementation**

I've made an additional improvement to the PR:

1. Completely removed the  function, as it was just a pass-through that didn't perform any transformation

2. Updated the relevant code to use model names as-is without any transformation

3. Simplified the tests to focus on the key functionality - proper handling of the  prefix

This makes the implementation cleaner and more maintainable, while still addressing the core issue of supporting the  prefix format.

cccntu commented:
**Minimal Changes for Maximum Effect**

I've further optimized this PR to include only the essential changes:

1. Completely reverted all changes to cost_calculator.py - leaving the original implementation intact

2. Only kept the small addition to get_llm_provider_logic.py to handle the vertex_ai/ prefix

3. Added tests to verify this minimal change works correctly

This approach addresses the core issue while touching as little code as possible, making it safer and easier to review/merge.

changjonathanc committed:
Add support for vertex_ai/ model name prefix format

This commit adds support for using the vertex_ai/ prefix with models in the format:
vertex_ai/claude-3-7-sonnet@20250219

The implementation:
1. Adds explicit handling for the vertex_ai/ prefix in get_llm_provider()
2. Correctly sets the custom_llm_provider to 'vertex_ai'
3. Returns the base model name without the prefix
4. Includes tests to verify the functionality

This allows for proper provider detection and cost calculation with Vertex AI hosted models.

cccntu commented:
**PR Updates: Squashed Into Single Clean Commit**

I have squashed all the commits into a single, clean commit that:

1. Adds explicit handling for the vertex_ai/ prefix in get_llm_provider()
2. Includes properly formatted tests with all lint issues fixed
3. Makes minimal changes to the codebase - just adding the prefix handling without modifying any existing behavior

This approach follows the pattern already established in the codebase for handling provider prefixes (e.g., azure/). All tests pass and the code fully conforms to the project's linting requirements.

The implementation preserves the original format of model names (including @ in date formats), ensuring proper cost calculation using the existing entries in the pricing database.

cccntu commented:
I think the main issue is this: 
https://github.com/BerriAI/litellm/issues/10181

closing this for now
